File: idakonyvtar/src/test/java/eu/eyan/idakonyvtar/testhelper/IdaLibraryTestHelper.java
package eu.eyan.idakonyvtar.testhelper;

import static org.fest.assertions.Assertions.assertThat;

import java.awt.Component;
import java.awt.Point;
import java.io.File;

import org.fest.swing.core.MouseButton;
import org.fest.swing.core.matcher.JButtonMatcher;
import org.fest.swing.fixture.DialogFixture;
import org.fest.swing.fixture.FrameFixture;

import eu.eyan.idakonyvtar.IdaLibrary;
import eu.eyan.idakonyvtar.text.LanguageHandler;
import eu.eyan.idakonyvtar.view.LibraryMenuAndToolBar;

public class IdaLibraryTestHelper {
	private FrameFixture frame;

	public void start(String filenév) {
		IdaLibrary.main(new String[] { filenév });
		frame = new FrameFixture(IdaLibrary.class.getName());
		frame.target.toFront();
	}

	public void requireVisible() {
		assertThat(frame.target.isVisible()).isTrue();
	}

	public void requireTitle(String title) {
		assertThat(frame.target.getTitle()).isEqualTo(title);
	}

	public void close() {
		frame.target.setVisible(false);
	}

	public void clickMenu(String menuPoint) {
		frame.menuItem(menuPoint).click();
	}

	public void load(File file) {
		clickMenu("File");
		clickMenu("Load");
		frame.fileChooser().selectFile(file).approve();
	}

	public void assertTableCell(int col, int row, String content) {
		assertThat(frame.table().contents()[row - 1][col - 1]).isEqualTo(content);
	}

	public void save(File file) {
		clickMenu("File");
		clickMenu("Save");
		frame.fileChooser().selectFile(file).approve();
	}

	public void clickNewButton() {
		frame.button(LibraryMenuAndToolBar.ADD_NEW_BOOK()).click();

	}

	public BookEditorTestHelper editor() {
		return new BookEditorTestHelper(frame.robot);
	}

	public void requireDeleteDisabled() {
		frame.button("Book törlése").requireDisabled();

	}

	public void selectRow(int row) {
		frame.table().selectRows(row - 1);
	}

	public void requireDeleteEnabled() {
		frame.button("Book törlése").requireEnabled();
	}

	public void clickDeleteButton() {
		frame.button("Book törlése").click();
	}

	public void clickApproveYes() {
		frame.dialog().button(JButtonMatcher.withText(LanguageHandler.YES())).click();
	}

	public void clickApproveNo() {
		frame.dialog().button(JButtonMatcher.withText(LanguageHandler.NO())).click();
	}

	public void filter(String filter) {
		frame.textBox(LibraryMenuAndToolBar.FILTER()).robot.enterText(filter);
	}

	public void assertTableRowCount(int rowsCount) {
		frame.table().requireRowCount(rowsCount);
	}

	public void doubleClick(int row) {
		int rowHeight = frame.table().target.getRowHeight();
		Point firstRow = frame.table().target.getLocationOnScreen();
		firstRow.translate(3, rowHeight * (row - 1) + rowHeight / 2);
		frame.table().robot.moveMouse(firstRow);
		frame.table().robot.click(firstRow, MouseButton.LEFT_BUTTON, 2);
	}

	public void cleanUp() {
		frame.cleanUp();
	}

	public void checkTitleWithNumber(int numberOfBooks) {
		if (numberOfBooks == 0)
			requireTitle("IdaLibrary - no books");
		else if (numberOfBooks == 1)
			requireTitle("IdaLibrary - one book");
		else
			requireTitle("IdaLibrary - " + numberOfBooks + " books");
	}

	public void clickExit() {
		frame.close();

	}

	public DialogFixture exitDialog() {
		return frame.dialog();
	}

	public void requireInvisible() {
		assertThat(frame.target.isVisible()).isFalse();
	}

	public Component getComponentToRecord() {
		return frame.target;
	}
}

File: idakonyvtar/src/test/scala/eu/eyan/idakonyvtar/testhelper/IdaLibraryTestHelper.scala
package eu.eyan.idakonyvtar.testhelper

import java.awt.Component
import java.awt.Point
import java.io.File

import org.fest.assertions.Assertions.assertThat
import org.fest.swing.core.MouseButton
import org.fest.swing.core.matcher.JButtonMatcher
import org.fest.swing.fixture.DialogFixture
import org.fest.swing.fixture.FrameFixture

import eu.eyan.idakonyvtar.IdaLibrary
import eu.eyan.idakonyvtar.text.TextsIda
import eu.eyan.testutil.TestPlus
import org.fest.swing.fixture.JMenuItemFixture
import eu.eyan.log.Log
import org.fest.swing.core.Robot
import org.fest.swing.core.BasicRobot
import org.fest.swing.exception.ComponentLookupException
import org.fest.swing.core.TypeMatcher
import java.awt.Dialog
import org.fest.swing.core.ComponentFoundCondition
import org.fest.swing.timing.Pause
import org.fest.swing.timing.Timeout

class IdaLibraryTestHelper extends TestPlus {

  var robot: Robot = null

  def dialog = {
    if (robot == null) robot = BasicRobot.robotWithCurrentAwtHierarchy
    val matcher = new TypeMatcher(classOf[Dialog], true)
    val description = "dialog to be found using matcher " + matcher
    val condition = new ComponentFoundCondition(description, robot.finder(), matcher)
    Pause.pause(condition, Timeout.timeout(100))
    new DialogFixture(robot, condition.found.asInstanceOf[Dialog])
  }

  def frame = {
    if (robot == null) robot = BasicRobot.robotWithCurrentAwtHierarchy
    new FrameFixture(robot, classOf[IdaLibrary].getName)
  }

  def cleanUp =
    if (robot != null) {
      robot.cleanUp
      pause(1000)
    }

  def start = { IdaLibrary.main(Array()); this }
  def start(filenév: String) = { IdaLibrary.main(Array(filenév)); this }

  def toFront = frame.target.toFront

  def requireVisible(): Unit = {
    assertThat(frame.target.isVisible).isTrue
  }

  def requireTitle(title: String): Unit = {
    assertThat(frame.target.getTitle).isEqualTo(title)
  }

  def close(): Unit = {
    frame.target.setVisible(false)
  }

  def clickMenu(menuPoint: String): Unit = {
    frame.menuItem(menuPoint).click()
  }

  implicit class JMenuItemFixtureImplicit(menuItem: JMenuItemFixture) {
    def requireText(expected: String) = menuItem.component.getText ==> expected
  }
  def requireMenuText(expected: String, path: String*) = menuItem(path: _*).requireText(expected)

  def menuItem(path: String*) = frame.menuItemWithPath(path: _*)

  def load(file: File): Unit = {
    clickMenu("File")
    clickMenu("Load")
    frame.fileChooser().selectFile(file).approve()
  }

  def assertTableCell(col: Int, row: Int, content: String) = assertThat(frame.table().contents()(row - 1)(col - 1)).isEqualTo(content)

  def save(file: File): Unit = {
    clickMenu("File")
    clickMenu("Save")
    frame.fileChooser().selectFile(file).approve()
  }

  def clickNewButton = frame.button("Új book hozzáadása").click()

  def editor = new BookEditorTestHelper(frame.robot)

  def requireDeleteDisabled(): Unit = {
    frame.button("Book törlése").requireDisabled()
  }

  def selectRow(row: Int): Unit = {
    frame.table().selectRows(row - 1)
  }

  def requireDeleteEnabled(): Unit = {
    frame.button("Book törlése").requireEnabled()
  }

  def clickDeleteButton(): Unit = {
    frame.button("Book törlése").click()
  }

  def clickApproveYes(): Unit = {
    frame
      .dialog()
      .button(JButtonMatcher.withText("Igen"))
      .click()
  }

  def clickApproveNo(): Unit = {
    frame
      .dialog()
      .button(JButtonMatcher.withText("Nem"))
      .click()
  }

  def filter(filter: String): Unit = {
    frame.textBox("filter").robot.enterText(filter)
  }

  def assertTableRowCount(rowsCount: Int): Unit = {
    frame.table().requireRowCount(rowsCount)
  }

  def doubleClick(row: Int): Unit = {
    val rowHeight: Int = frame.table().target.getRowHeight
    val firstRow: Point = frame.table().target.getLocationOnScreen
    firstRow.translate(3, rowHeight * (row - 1) + rowHeight / 2)
    frame.table().robot.moveMouse(firstRow)
    frame.table().robot.click(firstRow, MouseButton.LEFT_BUTTON, 2)
  }

  def checkTitleWithNumber(numberOfBooks: Int): Unit = {
    if (numberOfBooks == 0) requireTitle("IdaLibrary - no books")
    else if (numberOfBooks == 1) requireTitle("IdaLibrary - one book")
    else requireTitle("IdaLibrary - " + numberOfBooks + " books")
  }

  def clickExit = frame.close()

  def requireInvisible = assertThat(frame.target.isVisible).isFalse

  def requireNotExists = expectThrowable(classOf[ComponentLookupException], frame)

  def getComponentToRecord = frame.target
}
